File: hike_journal/domain/species_filters.py
from __future__ import annotations

from typing import Any


SPECIES_TYPE_OPTIONS = (
    "All types",
    "Plants",
    "Animals",
    "Birds",
    "Mammals",
    "Insects",
    "Reptiles",
    "Amphibians",
    "Arachnids",
    "Fungi",
    "Fish",
    "Mollusks",
    "Other life",
)

_ANIMAL_ICONIC_TAXA = {
    "animalia",
    "aves",
    "mammalia",
    "insecta",
    "reptilia",
    "amphibia",
    "arachnida",
    "actinopterygii",
    "mollusca",
}

_SPECIES_TYPE_TAXA = {
    "Plants": {"plantae"},
    "Animals": _ANIMAL_ICONIC_TAXA,
    "Birds": {"aves"},
    "Mammals": {"mammalia"},
    "Insects": {"insecta"},
    "Reptiles": {"reptilia"},
    "Amphibians": {"amphibia"},
    "Arachnids": {"arachnida"},
    "Fungi": {"fungi"},
    "Fish": {"actinopterygii"},
    "Mollusks": {"mollusca"},
}

_KNOWN_ICONIC_TAXA = {
    "plantae",
    "fungi",
    *_ANIMAL_ICONIC_TAXA,
}


def observation_matches_species_type(
    observation: dict[str, Any],
    species_type: str,
) -> bool:
    """Return whether an observation belongs in a Species Log type filter."""
    if species_type == "All types":
        return True
    iconic_taxon = str(observation.get("iconic_taxon_name") or "").strip().casefold()
    if species_type == "Other life":
        return iconic_taxon not in _KNOWN_ICONIC_TAXA
    return iconic_taxon in _SPECIES_TYPE_TAXA.get(species_type, set())

File: android/app/src/main/java/com/hikejournal/app/data/SpeciesFilters.kt
package com.hikejournal.app.data

enum class ObservationTypeFilter(val label: String) {
    All("All types"),
    Plants("Plants"),
    Animals("Animals"),
    Birds("Birds"),
    Mammals("Mammals"),
    Insects("Insects"),
    Reptiles("Reptiles"),
    Amphibians("Amphibians"),
    Arachnids("Arachnids"),
    Fungi("Fungi"),
    Fish("Fish"),
    Mollusks("Mollusks"),
    OtherLife("Other life"),
}

private val animalIconicTaxa = setOf(
    "animalia",
    "aves",
    "mammalia",
    "insecta",
    "reptilia",
    "amphibia",
    "arachnida",
    "actinopterygii",
    "mollusca",
)

private val knownIconicTaxa = animalIconicTaxa + setOf("plantae", "fungi")

fun iconicTaxonMatchesObservationType(
    iconicTaxonName: String,
    filter: ObservationTypeFilter,
): Boolean {
    val iconicTaxon = iconicTaxonName.trim().lowercase()
    return when (filter) {
        ObservationTypeFilter.All -> true
        ObservationTypeFilter.Plants -> iconicTaxon == "plantae"
        ObservationTypeFilter.Animals -> iconicTaxon in animalIconicTaxa
        ObservationTypeFilter.Birds -> iconicTaxon == "aves"
        ObservationTypeFilter.Mammals -> iconicTaxon == "mammalia"
        ObservationTypeFilter.Insects -> iconicTaxon == "insecta"
        ObservationTypeFilter.Reptiles -> iconicTaxon == "reptilia"
        ObservationTypeFilter.Amphibians -> iconicTaxon == "amphibia"
        ObservationTypeFilter.Arachnids -> iconicTaxon == "arachnida"
        ObservationTypeFilter.Fungi -> iconicTaxon == "fungi"
        ObservationTypeFilter.Fish -> iconicTaxon == "actinopterygii"
        ObservationTypeFilter.Mollusks -> iconicTaxon == "mollusca"
        ObservationTypeFilter.OtherLife -> iconicTaxon !in knownIconicTaxa
    }
}

fun SpeciesRecord.matchesObservationType(filter: ObservationTypeFilter): Boolean =
    iconicTaxonMatchesObservationType(iconicTaxonName, filter)

fun SpeciesRecord.matchesSpeciesSearch(query: String): Boolean {
    val normalizedQuery = query.trim()
    return normalizedQuery.isEmpty() ||
        commonName.contains(normalizedQuery, ignoreCase = true) ||
        scientificName.contains(normalizedQuery, ignoreCase = true) ||
        wikipediaSummary.contains(normalizedQuery, ignoreCase = true)
}

fun filterSpeciesBySearch(
    species: List<SpeciesRecord>,
    query: String,
): List<SpeciesRecord> =
    if (query.isBlank()) species else species.filter { it.matchesSpeciesSearch(query) }

fun filterSpeciesByObservationType(
    species: List<SpeciesRecord>,
    filter: ObservationTypeFilter,
): List<SpeciesRecord> =
    if (filter == ObservationTypeFilter.All) species
    else species.filter { it.matchesObservationType(filter) }
